sound/soc/sof/intel/hda-dsp.c

```diff
@@ -374,6 +378,12 @@ static int hda_resume(struct snd_sof_dev *sdev)
 	/* enable ppcap interrupt */
 	snd_hdac_ext_bus_ppcap_enable(bus, true);
 	snd_hdac_ext_bus_ppcap_int_enable(bus, true);
+
+	/*
+	 * power cycling of HDA codecs causes failures in HDA
+	 * controller logic -> force codecs to be powered
+	 */
+	snd_hda_set_power_save(hbus, -1);
```

Comment:
This seems a bit heavy-weight to me. In principle it would be enough to do snd_hda_set_power_save(hbus, SND_SOF_SUSPEND_DELAY_MS); once, and here just do some pm_runtime_get*() / put*(), similar to what is done in azx_vs_set_state():
```
			list_for_each_codec(codec, &chip->bus) {
				pm_runtime_suspend(hda_codec_dev(codec));
				pm_runtime_disable(hda_codec_dev(codec));
			}
...
			list_for_each_codec(codec, &chip->bus) {
				pm_runtime_enable(hda_codec_dev(codec));
				pm_runtime_resume(hda_codec_dev(codec));
			}
```
?

Reply:
@lyakh Thanks. That's a good comment. And in fact, this is probably needed. THe current iteration of the patch does not prevent resumes (only suspends) and it seems if I run stress test long enough, I can still trigger problems. Let me work on a update patch.

Reply:
@lyakh This proved harded, I tried pm_runtime_get(), pm_runtime_forbid(), pm_runtime_disable() and various combinations, but as this is called from within a suspend/resume, we hit issues.
I'll continue working on Monday. For now, the revert patch is the best cure.

Reply:
@kv2019i the order of runtime suspend/resume is dictated by the parent-child relationship of the devices. In our case, the sof-audio-pci dev is the parent of the hda codec device. Therefore, the sof device will be resumed prior to the codec and will be suspended after the codec. I'm not sure it is a good idea to mess with it.

Reply:
I tested your PR,  system hanged.